CODING.md

```diff
@@ -0,0 +1,97 @@
+- `Info` - Informative messages useful for node operators that do not indicate any fault or error.
+- `Debug` - Information concerning program logic decisions, diagnostic information, internal state, etc. which is primarily useful for developers.
+
+### Log Messages
+
+The key-value pairs (after a log message) should add more context to log messages. Keys are quite flexible and can contain more or less any string value. For consistency with existing code, there are a few conventions that should be followed:
+
+- Make your keys human-readable and be specific as much as possible (choose "peer_address" instead of "address" or "peer", "batch_id" instead of "id" or "batch", etc.)
+- Be consistent throughout the code base.
+- Use lower case for simple keys and lower_snake_case for more complex keys.
+
+Although key names are mostly unrestricted, it is generally a good idea to stick to printable ASCII characters, or at least to match the general character set of your log lines.
+
+`Error` and `Warning` log messages should provide enough information to identify the problem in the code base without the need to record the file name and line number alongside them, although this information can be added using the logging library.
+
+If you are writing/changing logs, it is recommended to look at the output. Answer questions like: is this what you expected? Is anything missing? Are they understandable? Do you need more/less context?
+
+### Users
+
+Two types of application users are identified: node operators and developers. Node operators should not be presented with confusing technical implementation details in log messages, but only with meaningful information regarding the functionality of the application in the form of meaningful statements. Developers are the users who know the implementation details and the technical details will help them in debugging the problematic state of the application.
+
+In general, the goal for developers is to create log messages that are easy to understand, analyze, and reference code so they can troubleshoot issues quickly. In the case of node operators, log messages should be concise and provide all the necessary information without unnecessary details.
+
+That is, the same problem event can have two log lines, but with different severity. This is the case for a combination of `Error/Debug` or `Warning/Debug`, where `Error` or `Warning` is for the node operator and `Debug` is for the developer to help them investigate the problem.
+
+> The `Error` level should be used with caution when it comes to node operators. It is recommended to log application code errors at the `Error` level if they require node operator intervention or if the node cannot continue its normal operation (unrecoverable errors). Otherwise, recoverable (application code) errors should be logged at the `Debug` level. The `Debug` level should not be used to monitor the node, but metrics should be used for this purpose.
+
```

Comment:
If the `Error` cannot be recovered from then I assume we also terminate the node in a timely manner after printing the `Error` log?

Reply:
Ideally yes (with the caveat that this comment is made with regards to logging only, not as to how the application itself should behave). I'm not sure if currently we panic on these sort of errors or if there's an actual shutdown mechanism (which is troublesome on its own when systemd or k8s are involved)